src/ucp/core/ucp_context.c

```diff
@@ -1234,9 +1226,9 @@ static ucs_status_t ucp_fill_resources(ucp_context_h context,
     /* Create memtype cache if we have memory type MDs, and it's enabled by
      * configuration
      */
-    if (context->num_mem_type_detect_mds && context->config.ext.enable_memtype_cache) {
-        status = ucs_memtype_cache_create(&context->memtype_cache);
-        if (status != UCS_OK) {
+    if (context->num_mem_type_detect_mds && ucs_global_opts.enable_memtype_cache) {
+        context->memtype_cache = ucs_memtype_cache_get_global();
+        if (context->memtype_cache == NULL) {
```

Comment:
i think there should not by any memtype_cache object in ucp/uct. ucs memtype cache functions should not accept any object as their parameter. they should operate directly on the global object.